x/cel/README.md

```diff
@@ -0,0 +1,98 @@
+# CEL module
+
+integrates cel-go
+
+## Introduction
+
+A cel-go expression type is defined by (EnvType, Input parameter type, Output parameter type).
+Expressions those shares the same functionality(e.g. a curve and its parameter updater) are under the same EnvType.
+We call a set of expressions a cell. Expressions under a cell should share same EnvType(which acts like a state). 
+Expressions can have different input/output parameter types. Output type includes effects.
+For the sake of simplicity, we read the entire variable set defined by EnvType. This could be optimized later.
+For this reason we allow finite set of variables used inside the expression.
+Later we can parameterize EnvSet to use specific variable only.
+
+## Components
+
+### EnvType
+
+An EnvType is, roughly, a JSON object type, such as:
+
+```ts
+interface User {
+  balance: number;
+  moniker: string;
+  minter:  address;
+}
+```
+
+An EnvType defines the state structure that is going to be used by a set of expressions.
+When executing an expression, the module will load corresponding EnvType from the state,
+and initializes a cel-go Env providing the EnvSet declarations to cel.NewEnv.
+
+Custom functions(such as precompiled curve calculation) is EnvType-dependent. 
+For Osmosis, my opinion is that EnvType extension should be done on chain level governance.
+
+### Cell
+
+A cell is somewhat like a contract. Cell can consists of multiple expressions, which shares the same state.
+New expressions can be dynamically added / existing expressions can be updated through cell-level governance.
+
+### Expression
+
+Three types of expressions:
+1. Input -> Output (pure function)
```

Comment:
Have you had a chance to think about how we would go about routing state entries to inputs?

We want to allow other entries in state to be inputs to a cel AMM expression, but I'm not really sure how the abstraction for this should work.

Should this be handled at the expression level, or at a higher level for defining your Cel program. Handling at the level of defining the entire cel program would allow you to read it, and then alias it as a normal input for the rest of the program (and ditto for the output).

Reply:
Cel allows the user to input arbitrary variables as env parameters. Take an example state:

```go
type ExampleState struct {
  id uint64
  name string
}
```

In this case, the cel module can read the state from the KVStore, and route the state as Env parameter `{"id": "100", "name": "examplename"}`. 

Cel expressions do not have explicit function forms, but the variables inside the expressions have assigned values in their execution environment, so there is no distinction between cel module provided vars and external module provided vars. So as you said, the state read will be treated as the same as accessing function parameters(in this point, I made a mistake, and type 1 and 2 do not have to be separated).

The output is a bit more complex - we might not want to simply have `State -> State` transition model, because there could be more side effect actions that an expression wants to request to the module. I will writeup about this.